### PR TITLE
KUBSAN: Add nearly all missing -fsanitize handlers

### PR DESCRIPTION
**Pull request by Hendiadyoin1**
A step to add more ubsan support in the kernel.

Missing handlers:
* cfi_bad_type
* cf_bad_icall
* cfi_chack_fail

Some use of the "ValueHandel"s is included, but with or with out boot crashes

Build succeeds at the moment, but boot fails to early for backtracing:

<details>
<summary>Error output</summary>

```
Executing task in folder serenity: qemu-system-i386 -L "C:\Program Files\qemu" -s -m 4G -cpu max -d cpu_reset,guest_errors -no-reboot -smp 2 -device VGA,vgamem_mb=64 -drive file=_disk_image,format=raw,index=0,media=disk -device ich9-ahci -usb -debugcon stdio -soundhw pcspk -device sb16 -netdev user,id=breh,hostfwd=tcp:127.0.0.1:8888-10.0.2.15:8888,hostfwd=tcp:127.0.0.1:8823-10.0.2.15:23 -device e1000,netdev=breh -kernel Kernel/Kernel -append hello <

CPU Reset (CPU 0)
EAX=00000000 EBX=00000000 ECX=00000000 EDX=00000000
ESI=00000000 EDI=00000000 EBP=00000000 ESP=00000000
EIP=00000000 EFL=00000000 [-------] CPL=0 II=0 A20=0 SMM=0 HLT=0
ES =0000 00000000 00000000 00000000
CS =0000 00000000 00000000 00000000
SS =0000 00000000 00000000 00000000
DS =0000 00000000 00000000 00000000
FS =0000 00000000 00000000 00000000
GS =0000 00000000 00000000 00000000
LDT=0000 00000000 00000000 00000000
TR =0000 00000000 00000000 00000000
GDT=     00000000 00000000
IDT=     00000000 00000000
CR0=00000000 CR2=00000000 CR3=00000000 CR4=00000000
DR0=00000000 DR1=00000000 DR2=00000000 DR3=00000000
DR6=00000000 DR7=00000000
CCS=00000000 CCD=00000000 CCO=DYNAMIC
EFER=0000000000000000
FCW=0000 FSW=0000 [ST=0] FTW=ff MXCSR=00000000
FPR0=0000000000000000 0000 FPR1=0000000000000000 0000
FPR2=0000000000000000 0000 FPR3=0000000000000000 0000
FPR4=0000000000000000 0000 FPR5=0000000000000000 0000
FPR6=0000000000000000 0000 FPR7=0000000000000000 0000
XMM00=00000000000000000000000000000000 XMM01=00000000000000000000000000000000
XMM02=00000000000000000000000000000000 XMM03=00000000000000000000000000000000
XMM04=00000000000000000000000000000000 XMM05=00000000000000000000000000000000
XMM06=00000000000000000000000000000000 XMM07=00000000000000000000000000000000
CPU Reset (CPU 1)
EAX=00000000 EBX=00000000 ECX=00000000 EDX=00000000
ESI=00000000 EDI=00000000 EBP=00000000 ESP=00000000
EIP=00000000 EFL=00000000 [-------] CPL=0 II=0 A20=0 SMM=0 HLT=0
ES =0000 00000000 00000000 00000000
CS =0000 00000000 00000000 00000000
SS =0000 00000000 00000000 00000000
DS =0000 00000000 00000000 00000000
FS =0000 00000000 00000000 00000000
GS =0000 00000000 00000000 00000000
LDT=0000 00000000 00000000 00000000
TR =0000 00000000 00000000 00000000
GDT=     00000000 00000000
IDT=     00000000 00000000
CR0=00000000 CR2=00000000 CR3=00000000 CR4=00000000
DR0=00000000 DR1=00000000 DR2=00000000 DR3=00000000
DR6=00000000 DR7=00000000
CCS=00000000 CCD=00000000 CCO=DYNAMIC
EFER=0000000000000000
FCW=0000 FSW=0000 [ST=0] FTW=ff MXCSR=00000000
FPR0=0000000000000000 0000 FPR1=0000000000000000 0000
FPR2=0000000000000000 0000 FPR3=0000000000000000 0000
FPR4=0000000000000000 0000 FPR5=0000000000000000 0000
FPR6=0000000000000000 0000 FPR7=0000000000000000 0000
XMM00=00000000000000000000000000000000 XMM01=00000000000000000000000000000000
XMM02=00000000000000000000000000000000 XMM03=00000000000000000000000000000000
XMM04=00000000000000000000000000000000 XMM05=00000000000000000000000000000000
XMM06=00000000000000000000000000000000 XMM07=00000000000000000000000000000000
qemu-system-i386: multiboot knows VBE. we don't
qemu-system-i386: warning: '-soundhw pcspk' is deprecated, please set a backend using '-machine pcspk-audiodev=<name>' instead
CPU Reset (CPU 0)
EAX=00000000 EBX=00000000 ECX=00000000 EDX=00000663
ESI=00000000 EDI=00000000 EBP=00000000 ESP=00000000
EIP=0000fff0 EFL=00000002 [-------] CPL=0 II=0 A20=1 SMM=0 HLT=0
ES =0000 00000000 0000ffff 00009300
CS =f000 ffff0000 0000ffff 00009b00
SS =0000 00000000 0000ffff 00009300
DS =0000 00000000 0000ffff 00009300
FS =0000 00000000 0000ffff 00009300
GS =0000 00000000 0000ffff 00009300
LDT=0000 00000000 0000ffff 00008200
TR =0000 00000000 0000ffff 00008b00
GDT=     00000000 0000ffff
IDT=     00000000 0000ffff
CR0=60000010 CR2=00000000 CR3=00000000 CR4=00000000
DR0=00000000 DR1=00000000 DR2=00000000 DR3=00000000
DR6=ffff0ff0 DR7=00000400
CCS=00000000 CCD=00000000 CCO=DYNAMIC
EFER=0000000000000000
FCW=037f FSW=0000 [ST=0] FTW=00 MXCSR=00001f80
FPR0=0000000000000000 0000 FPR1=0000000000000000 0000
FPR2=0000000000000000 0000 FPR3=0000000000000000 0000
FPR4=0000000000000000 0000 FPR5=0000000000000000 0000
FPR6=0000000000000000 0000 FPR7=0000000000000000 0000
XMM00=00000000000000000000000000000000 XMM01=00000000000000000000000000000000
XMM02=00000000000000000000000000000000 XMM03=00000000000000000000000000000000
XMM04=00000000000000000000000000000000 XMM05=00000000000000000000000000000000
XMM06=00000000000000000000000000000000 XMM07=00000000000000000000000000000000
CPU Reset (CPU 1)
EAX=00000000 EBX=00000000 ECX=00000000 EDX=00000663
ESI=00000000 EDI=00000000 EBP=00000000 ESP=00000000
EIP=0000fff0 EFL=00000002 [-------] CPL=0 II=0 A20=1 SMM=0 HLT=1
ES =0000 00000000 0000ffff 00009300
CS =f000 ffff0000 0000ffff 00009b00
SS =0000 00000000 0000ffff 00009300
DS =0000 00000000 0000ffff 00009300
FS =0000 00000000 0000ffff 00009300
GS =0000 00000000 0000ffff 00009300
LDT=0000 00000000 0000ffff 00008200
TR =0000 00000000 0000ffff 00008b00
GDT=     00000000 0000ffff
IDT=     00000000 0000ffff
CR0=60000010 CR2=00000000 CR3=00000000 CR4=00000000
DR0=00000000 DR1=00000000 DR2=00000000 DR3=00000000
DR6=ffff0ff0 DR7=00000400
CCS=00000000 CCD=00000000 CCO=DYNAMIC
EFER=0000000000000000
FCW=037f FSW=0000 [ST=0] FTW=00 MXCSR=00001f80
FPR0=0000000000000000 0000 FPR1=0000000000000000 0000
FPR2=0000000000000000 0000 FPR3=0000000000000000 0000
FPR4=0000000000000000 0000 FPR5=0000000000000000 0000
FPR6=0000000000000000 0000 FPR7=0000000000000000 0000
XMM00=00000000000000000000000000000000 XMM01=00000000000000000000000000000000
XMM02=00000000000000000000000000000000 XMM03=00000000000000000000000000000000
XMM04=00000000000000000000000000000000 XMM05=00000000000000000000000000000000
XMM06=00000000000000000000000000000000 XMM07=00000000000000000000000000000000
CPU Reset (CPU 1)
EAX=00000000 EBX=00000000 ECX=00000000 EDX=00000663
ESI=00000000 EDI=00000000 EBP=00000000 ESP=00000000
EIP=0000fff0 EFL=00000002 [-------] CPL=0 II=0 A20=1 SMM=0 HLT=0
ES =0000 00000000 0000ffff 00009300
CS =f000 ffff0000 0000ffff 00009b00
SS =0000 00000000 0000ffff 00009300
DS =0000 00000000 0000ffff 00009300
FS =0000 00000000 0000ffff 00009300
GS =0000 00000000 0000ffff 00009300
LDT=0000 00000000 0000ffff 00008200
TR =0000 00000000 0000ffff 00008b00
GDT=     00000000 0000ffff
IDT=     00000000 0000ffff
CR0=60000010 CR2=00000000 CR3=00000000 CR4=00000000
DR0=00000000 DR1=00000000 DR2=00000000 DR3=00000000
DR6=ffff0ff0 DR7=00000400
CCS=00000000 CCD=00000000 CCO=EFLAGS
EFER=0000000000000000
FCW=037f FSW=0000 [ST=0] FTW=00 MXCSR=00001f80
FPR0=0000000000000000 0000 FPR1=0000000000000000 0000
FPR2=0000000000000000 0000 FPR3=0000000000000000 0000
FPR4=0000000000000000 0000 FPR5=0000000000000000 0000
FPR6=0000000000000000 0000 FPR7=0000000000000000 0000
XMM00=00000000000000000000000000000000 XMM01=00000000000000000000000000000000
XMM02=00000000000000000000000000000000 XMM03=00000000000000000000000000000000
XMM04=00000000000000000000000000000000 XMM05=00000000000000000000000000000000
XMM06=00000000000000000000000000000000 XMM07=00000000000000000000000000000000
Triple fault
```

</details>

### CODE REVIEW

**Comment by Lubrsi** [comment]
> Build succeeds at the moment, but boot fails to early for backtracing:

I believe this is caused by the pointer overflow check. I tried implementing that check myself but it did the same boot looping you're seeing.

**Comment by Hendiadyoin1** [comment]
After looking into wich flags cause a crash, I found every flag in which I was using the value handles.

I suppose its something with access rights (in ring3?)
In the gcc implementation they first create a copy of the source, maybe that will solve the issue

**Comment by Hendiadyoin1** [comment]
Also ignore the lagom build errors they are caused by something in the masterbranch at the time of the fork

**Comment by Hendiadyoin1** [comment]
Apparently the pure existance of some of these handles causes a triple fault, with no amount of debugability.

With that said all handles except the aforementioned cfi handles are implemented.  
The only flags not working are:
* null
* alignment
* pointer-overflow

**Comment by awesomekling** [comment]
The build is failing because you need to use `clang-format` to format the source code. :)

**Comment by Hendiadyoin1** [comment]
Formatting is fixed now, the only thing that might be bugging someone is  that the build a quite verbose about the fsanitize flags being used

File: Kernel/UBSanitizer.cpp [hunk 1]
```diff
@@ -53,73 +58,156 @@ void __ubsan_handle_nonnull_arg(const NonnullArgData& data)
...
+}
+
+void __ubsan_handle_invalid_builtin(const InvalidBuiltinData);
+void __ubsan_handle_invalid_builtin(const InvalidBuiltinData data)
+{
+    dbgln("KUBSAN: passing invalid argument");
+    print_location(data.location);
+}
+
+// FIXME: Causes a triple fault on boot
```
Comment (bgianfo) [inline Kernel/UBSanitizer.cpp:198]: Is it a trouble fault? I tried adding this previously and when I stepped through in gdb it looked like it was faulting on the `movl %ebx, multiboot_info_ptr` instruction before we even enter any C++ code? 
``` asm
    /* jump into C++ land */
    addl $0xc0000000, %ebx
    movl %ebx, multiboot_info_ptr


    call init
    add $4, %esp
```

Comment (Hendiadyoin1) [reply Kernel/UBSanitizer.cpp:198]: Im not using any debug tools and the only output I get is the following (similar to the one mentioned above):

<details>
<summary>Build and Error Log</summary>

```
> Executing task: wsl bash -c "cd Build && ninja install && ninja image" <

[0/2] Re-checking globbed directories...
[1/2] Re-running CMake...
        kernel fsanitize flag:shift
        kernel fsanitize flag:shift-exponent
        kernel fsanitize flag:shift-base
        kernel fsanitize flag:integer-divide-by-zero
        kernel fsanitize flag:unreachable
        kernel fsanitize flag:vla-bound
        kernel fsanitize flag:return
        kernel fsanitize flag:signed-integer-overflow
        kernel fsanitize flag:bounds
        kernel fsanitize flag:bounds-strict
        kernel fsanitize flag:object-size
        kernel fsanitize flag:float-divide-by-zero
        kernel fsanitize flag:float-cast-overflow
        kernel fsanitize flag:nonnull-attribute
        kernel fsanitize flag:returns-nonnull-attribute
        kernel fsanitize flag:bool
        kernel fsanitize flag:enum
        kernel fsanitize flag:vptr
        kernel fsanitize flag:builtin
        kernel fsanitize flag:pointer-overflow
CMake Warning at Meta/CMake/utils.cmake:89 (message):
  Missing medium app icon:
  /mnt/serenity/Base/res/icons/32x32/audio-volume-high.png
Call Stack (most recent call first):
  Userland/MenuApplets/Audio/CMakeLists.txt:5 (serenity_app)


CMake Warning at Meta/CMake/utils.cmake:89 (message):
  Missing medium app icon: /mnt/serenity/Base/res/icons/32x32/clipboard.png
Call Stack (most recent call first):
  Userland/MenuApplets/ClipboardHistory/CMakeLists.txt:6 (serenity_app)


CMake Warning at Meta/CMake/utils.cmake:89 (message):
  Missing medium app icon: /mnt/serenity/Base/res/icons/32x32/network.png
Call Stack (most recent call first):
  Userland/MenuApplets/Network/CMakeLists.txt:5 (serenity_app)


-- Configuring done
-- Generating done
-- Build files have been written to: /mnt/serenity/Build
[0/2] Re-checking globbed directories...
[240/241] Install the project...
-- Install configuration: ""
[0/2] Re-checking globbed directories...
[0/1] cd /mnt/serenity/Build && /usr/bin/cmake -E env SERENITY_ROOT=/mnt/serenity SERENITY_ARCH=i686 /mnt/serenity/Meta/build-image-qemu.sh
setting up disk image...
Formatting '_disk_image', fmt=raw size=222298112
done
creating new filesystem... done
mounting filesystem... mount: mnt/: mount failed: Operation not permitted.
mount failed but genext2fs exists, use it instead
done
installing base system... done
creating initial filesystem structure... done
creating utmp file... done
setting up device nodes folder... done
writing version file... done
installing users... done
adding some desktop icons...done
installing shortcuts... done
installing 'checksum' variants... done
done

Terminal will be reused by tasks, press any key to close it.

> Executing task: qemu-system-i386 -L "C:\Program Files\qemu" -s -m 4G -cpu max -d cpu_reset,guest_errors -no-reboot -smp 2 -device VGA,vgamem_mb=64 -drive file=_disk_image,format=raw,index=0,media=disk -device ich9-ahci -usb -debugcon stdio -soundhw pcspk -device sb16 -netdev user,id=breh,hostfwd=tcp:127.0.0.1:8888-10.0.2.15:8888,hostfwd=tcp:127.0.0.1:8823-10.0.2.15:23 -device e1000,netdev=breh -kernel Kernel/Kernel -append hello <

CPU Reset (CPU 0)
EAX=00000000 EBX=00000000 ECX=00000000 EDX=00000000
ESI=00000000 EDI=00000000 EBP=00000000 ESP=00000000
EIP=00000000 EFL=00000000 [-------] CPL=0 II=0 A20=0 SMM=0 HLT=0
ES =0000 00000000 00000000 00000000
CS =0000 00000000 00000000 00000000
SS =0000 00000000 00000000 00000000
DS =0000 00000000 00000000 00000000
FS =0000 00000000 00000000 00000000
GS =0000 00000000 00000000 00000000
LDT=0000 00000000 00000000 00000000
TR =0000 00000000 00000000 00000000
GDT=     00000000 00000000
IDT=     00000000 00000000
CR0=00000000 CR2=00000000 CR3=00000000 CR4=00000000
DR0=00000000 DR1=00000000 DR2=00000000 DR3=00000000 
DR6=00000000 DR7=00000000
CCS=00000000 CCD=00000000 CCO=DYNAMIC
EFER=0000000000000000
FCW=0000 FSW=0000 [ST=0] FTW=ff MXCSR=00000000
FPR0=0000000000000000 0000 FPR1=0000000000000000 0000
FPR2=0000000000000000 0000 FPR3=0000000000000000 0000
FPR4=0000000000000000 0000 FPR5=0000000000000000 0000
FPR6=0000000000000000 0000 FPR7=0000000000000000 0000
XMM00=00000000000000000000000000000000 XMM01=00000000000000000000000000000000
XMM02=00000000000000000000000000000000 XMM03=00000000000000000000000000000000
XMM04=00000000000000000000000000000000 XMM05=00000000000000000000000000000000
XMM06=00000000000000000000000000000000 XMM07=00000000000000000000000000000000
CPU Reset (CPU 1)
EAX=00000000 EBX=00000000 ECX=00000000 EDX=00000000
ESI=00000000 EDI=00000000 EBP=00000000 ESP=00000000
EIP=00000000 EFL=00000000 [-------] CPL=0 II=0 A20=0 SMM=0 HLT=0
ES =0000 00000000 00000000 00000000
CS =0000 00000000 00000000 00000000
SS =0000 00000000 00000000 00000000
DS =0000 00000000 00000000 00000000
FS =0000 00000000 00000000 00000000
GS =0000 00000000 00000000 00000000
LDT=0000 00000000 00000000 00000000
TR =0000 00000000 00000000 00000000
GDT=     00000000 00000000
IDT=     00000000 00000000
CR0=00000000 CR2=00000000 CR3=00000000 CR4=00000000
DR0=00000000 DR1=00000000 DR2=00000000 DR3=00000000
DR6=00000000 DR7=00000000
CCS=00000000 CCD=00000000 CCO=DYNAMIC 
EFER=0000000000000000
FCW=0000 FSW=0000 [ST=0] FTW=ff MXCSR=00000000
FPR0=0000000000000000 0000 FPR1=0000000000000000 0000
FPR2=0000000000000000 0000 FPR3=0000000000000000 0000
FPR4=0000000000000000 0000 FPR5=0000000000000000 0000
FPR6=0000000000000000 0000 FPR7=0000000000000000 0000
XMM00=00000000000000000000000000000000 XMM01=00000000000000000000000000000000
XMM02=00000000000000000000000000000000 XMM03=00000000000000000000000000000000
XMM04=00000000000000000000000000000000 XMM05=00000000000000000000000000000000
XMM06=00000000000000000000000000000000 XMM07=00000000000000000000000000000000
qemu-system-i386: multiboot knows VBE. we don't
qemu-system-i386: warning: '-soundhw pcspk' is deprecated, please set a backend using '-machine pcspk-audiodev=<name>' instead
CPU Reset (CPU 0)
EAX=00000000 EBX=00000000 ECX=00000000 EDX=00000663
ESI=00000000 EDI=00000000 EBP=00000000 ESP=00000000
EIP=0000fff0 EFL=00000002 [-------] CPL=0 II=0 A20=1 SMM=0 HLT=0
ES =0000 00000000 0000ffff 00009300
CS =f000 ffff0000 0000ffff 00009b00
SS =0000 00000000 0000ffff 00009300
DS =0000 00000000 0000ffff 00009300
FS =0000 00000000 0000ffff 00009300
GS =0000 00000000 0000ffff 00009300
LDT=0000 00000000 0000ffff 00008200
TR =0000 00000000 0000ffff 00008b00
GDT=     00000000 0000ffff
IDT=     00000000 0000ffff
CR0=60000010 CR2=00000000 CR3=00000000 CR4=00000000
DR0=00000000 DR1=00000000 DR2=00000000 DR3=00000000
DR6=ffff0ff0 DR7=00000400
CCS=00000000 CCD=00000000 CCO=DYNAMIC
EFER=0000000000000000
FCW=037f FSW=0000 [ST=0] FTW=00 MXCSR=00001f80
FPR0=0000000000000000 0000 FPR1=0000000000000000 0000
FPR2=0000000000000000 0000 FPR3=0000000000000000 0000
FPR4=0000000000000000 0000 FPR5=0000000000000000 0000
FPR6=0000000000000000 0000 FPR7=0000000000000000 0000
XMM00=00000000000000000000000000000000 XMM01=00000000000000000000000000000000
XMM02=00000000000000000000000000000000 XMM03=00000000000000000000000000000000
XMM04=00000000000000000000000000000000 XMM05=00000000000000000000000000000000
XMM06=00000000000000000000000000000000 XMM07=00000000000000000000000000000000
CPU Reset (CPU 1)
EAX=00000000 EBX=00000000 ECX=00000000 EDX=00000663
ESI=00000000 EDI=00000000 EBP=00000000 ESP=00000000
EIP=0000fff0 EFL=00000002 [-------] CPL=0 II=0 A20=1 SMM=0 HLT=1
ES =0000 00000000 0000ffff 00009300
CS =f000 ffff0000 0000ffff 00009b00
SS =0000 00000000 0000ffff 00009300
DS =0000 00000000 0000ffff 00009300
FS =0000 00000000 0000ffff 00009300
GS =0000 00000000 0000ffff 00009300
LDT=0000 00000000 0000ffff 00008200
TR =0000 00000000 0000ffff 00008b00
GDT=     00000000 0000ffff
IDT=     00000000 0000ffff
CR0=60000010 CR2=00000000 CR3=00000000 CR4=00000000
DR0=00000000 DR1=00000000 DR2=00000000 DR3=00000000
DR6=ffff0ff0 DR7=00000400
CCS=00000000 CCD=00000000 CCO=DYNAMIC 
EFER=0000000000000000
FCW=037f FSW=0000 [ST=0] FTW=00 MXCSR=00001f80
FPR0=0000000000000000 0000 FPR1=0000000000000000 0000
FPR2=0000000000000000 0000 FPR3=0000000000000000 0000
FPR4=0000000000000000 0000 FPR5=0000000000000000 0000
FPR6=0000000000000000 0000 FPR7=0000000000000000 0000
XMM00=00000000000000000000000000000000 XMM01=00000000000000000000000000000000
XMM02=00000000000000000000000000000000 XMM03=00000000000000000000000000000000
XMM04=00000000000000000000000000000000 XMM05=00000000000000000000000000000000
XMM06=00000000000000000000000000000000 XMM07=00000000000000000000000000000000
CPU Reset (CPU 1)
EAX=00000000 EBX=00000000 ECX=00000000 EDX=00000663
ESI=00000000 EDI=00000000 EBP=00000000 ESP=00000000
EIP=0000fff0 EFL=00000002 [-------] CPL=0 II=0 A20=1 SMM=0 HLT=0
ES =0000 00000000 0000ffff 00009300
CS =f000 ffff0000 0000ffff 00009b00
SS =0000 00000000 0000ffff 00009300
DS =0000 00000000 0000ffff 00009300
FS =0000 00000000 0000ffff 00009300
GS =0000 00000000 0000ffff 00009300
LDT=0000 00000000 0000ffff 00008200
TR =0000 00000000 0000ffff 00008b00
GDT=     00000000 0000ffff
IDT=     00000000 0000ffff
CR0=60000010 CR2=00000000 CR3=00000000 CR4=00000000
DR0=00000000 DR1=00000000 DR2=00000000 DR3=00000000
DR6=ffff0ff0 DR7=00000400
CCS=00000000 CCD=00000000 CCO=EFLAGS  
EFER=0000000000000000
FCW=037f FSW=0000 [ST=0] FTW=00 MXCSR=00001f80
FPR0=0000000000000000 0000 FPR1=0000000000000000 0000
FPR2=0000000000000000 0000 FPR3=0000000000000000 0000
FPR4=0000000000000000 0000 FPR5=0000000000000000 0000
FPR6=0000000000000000 0000 FPR7=0000000000000000 0000
XMM00=00000000000000000000000000000000 XMM01=00000000000000000000000000000000
XMM02=00000000000000000000000000000000 XMM03=00000000000000000000000000000000
XMM04=00000000000000000000000000000000 XMM05=00000000000000000000000000000000
XMM06=00000000000000000000000000000000 XMM07=00000000000000000000000000000000
Triple fault
```

</details>

As you can see the last thing qemu tells me is that there was a tripple fault and the proccessor state at that time.  
If I'm reading it correctly the instruction pointer is 0x0000fff0 which doesnt look that good.

You get pretty much the same stacktrace when trying any of the other two failing flags.  

As you said the fault probably happens before any C++ generated code is run, maybe that is the problem, that gcc tries to audit something before this level of auditing is even possible, like trying to call the handlers without even having a stack or in the wrong bit-mode.

Comment (Hendiadyoin1) [reply Kernel/UBSanitizer.cpp:198]: Is it possible, that ebx is 0 at that point in time and we add 0xc0000000 to 0?  
I havent found any other mention of ebx in that file.

My guess would there fore be:
```asm
/* jump into C++ land */
    movl $0xc0000000, %ebx
    movl %ebx, multiboot_info_ptr


    call init
    add $4, %esp
```

but with that I still get the same error

Comment (Hendiadyoin1) [reply Kernel/UBSanitizer.cpp:198]: After stepping through with gdb I can say with confidence, that it does not crash at this point in time

**Comment by Hendiadyoin1** [comment]
Apparently`strlen(const char* str)` is the culprit for the pointer overflow, causing a jump into the stack and after that a pagefault, that causes another pagefault/doublefault wich causes just another pagefault/triplefault.

[fault_log.txt](https://github.com/SerenityOS/serenity/files/5952174/fault_log.txt)
The first error appears occurs at or after the execution of `0xc01e0606` which is the escape condition of strlen:
```asm
c01e05ca <strlen>:
c01e05ca:       55                      push   %ebp
c01e05cb:       31 c0                   xor    %eax,%eax
c01e05cd:       89 e5                   mov    %esp,%ebp
c01e05cf:       57                      push   %edi
c01e05d0:       56                      push   %esi
c01e05d1:       53                      push   %ebx
c01e05d2:       e8 26 37 0c 00          call   c02a3cfd <__x86.get_pc_thunk.bx>
c01e05d7:       81 c3 c5 4b f3 ff       add    $0xfff34bc5,%ebx
c01e05dd:       83 ec 1c                sub    $0x1c,%esp
c01e05e0:       8b 75 08                mov    0x8(%ebp),%esi
c01e05e3:       8d bb a8 e7 28 00       lea    0x28e7a8(%ebx),%edi
c01e05e9:       8d 14 06                lea    (%esi,%eax,1),%edx
c01e05ec:       8d 4a 01                lea    0x1(%edx),%ecx
c01e05ef:       83 fa ff                cmp    $0xffffffff,%edx
c01e05f2:       72 12                   jb     c01e0606 <strlen+0x3c>
c01e05f4:       89 45 e4                mov    %eax,-0x1c(%ebp)
c01e05f7:       50                      push   %eax
c01e05f8:       51                      push   %ecx
c01e05f9:       52                      push   %edx
c01e05fa:       57                      push   %edi
c01e05fb:       e8 e8 e4 05 00          call   c023eae8 <__ubsan_handle_pointer_overflow>
c01e0600:       8b 45 e4                mov    -0x1c(%ebp),%eax
c01e0603:       83 c4 10                add    $0x10,%esp
c01e0606:       80 3c 06 00             cmpb   $0x0,(%esi,%eax,1)
c01e060a:       74 03                   je     c01e060f <strlen+0x45>
c01e060c:       40                      inc    %eax
c01e060d:       eb da                   jmp    c01e05e9 <strlen+0x1f>
c01e060f:       8d 65 f4                lea    -0xc(%ebp),%esp
c01e0612:       5b                      pop    %ebx
c01e0613:       5e                      pop    %esi
c01e0614:       5f                      pop    %edi
```

**Comment by Hendiadyoin1** [comment]
The tripple fault seems to be caused, because the fault handling isnt initialized yet.  
When moving IDT initialization (Processor::early_init->HeapInit->Processor::init) to the front of init it gives following output and then loops probably somwhere in `smp_process_pending_messages()`, I guess because I havent initialized the command line, but that relies on strlen, which the part that breaks.

Could someone else try to reorganize init() who has more knowledge of the dependencies to get it to run with the pointer-overflow flag, because honestly I have no clue what I am supposed to do to get some more debug output other than taking apart some of the initialization logic and making the init function a mess

**Comment by awesomekling** [comment]
I think we can merge the first steps here and then spin off the remaining weird crashes in a separate PR.

But first, please squash the fixups into the commits they are fixing instead of adding them as separate commits :)

**Comment by Hendiadyoin1** [comment]
> But first, please squash the fixups into the commits they are fixing instead of adding them as separate commits :)

Sadly I have no clue how to edit the git history on github, as you may see I've tried.
If you want you can merge, or if you want a cleaner history I can open a new fork with a clean history

**Comment by tomuta** [comment]
> > But first, please squash the fixups into the commits they are fixing instead of adding them as separate commits :)
> 
> Sadly I have no clue how to edit the git history on github, as you may see I've tried.
> If you want you can merge, or if you want a cleaner history I can open a new fork with a clean history

It's a fairly easy process. Just edit/swash commits using `git rebase -i` and then when you're done just force-push your modified branch again using `git push --force`. It's the same process, github doesn't limit you in any way.

**Comment by Hendiadyoin1** [comment]
It's just one big commmit now, because if essentially done only one thing

**Comment by awesomekling** [comment]
Excellent work! Thank you :)

File: Kernel/UBSanitizer.cpp (hunk 1 above)
Comment (Hendiadyoin1) [reply Kernel/UBSanitizer.cpp:198]: Oh ive misread your question, you did not mean the pointer-overflow flag but the allignment flag, and then yes it does pagefault (I really dont know how, but the error code I see is 0xe wich is a pagefault) there, but there arent any faulthandlers regustered at that point therefore the fault escalates to a triple fault

**Pull request merged**